File: 3/3.cpp
#include <iostream>
#include <math.h>


bool isPrime(long N) {
  int limit = (int) ceil(sqrt(N));
  for (int i=2; i<limit+1; i++) {
    if (N % i == 0) {
      return false;
    }
  }
  return true;
}

int largestPrimeFactor(long N) {
  int limit = (int) ceil(sqrt(N));
  int largest = 0;
  for (int i=limit; i>1; i--) {
    if (N % i == 0) {
      if (isPrime(i) and i > largest) {
        largest = i;
      }
      if (isPrime(N/i) and N/i > largest) {
        largest = N/i;
      }
    }
  }
  return largest;
}


int main() {
  int result = largestPrimeFactor(600851475143);
  std::cout << result;
}

File: 3/3.py
from math import sqrt

def is_prime(n):
    limit = int(sqrt(n)) + 1
    for x in xrange(2, limit+1):
        if n % x == 0:
            return False
    return True

def main(n):
    pfs = []
    limit = int(sqrt(n)) + 1
    for x in xrange(limit+1, 1, -1):
        if n % x == 0:
            if is_prime(x):
                pfs.append(x)
            if n/x != x:
                if is_prime(n/x):
                    pfs.append(n/x)
    return max(pfs)


print main(600851475143)
